Services.Test/CustomDeviceModelsTest.cs

```diff
@@ -0,0 +1,342 @@
+            this.UpdateDeviceModelInStorage(deviceModel);
+
+            // Act
+            this.target.InsertAsync(deviceModel).Wait(TimeSpan.FromSeconds(30));
+
+            // Assert
+            this.storage.Verify(x => x.UpdateAsync(
+                STORAGE_COLLECTION,
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                null), Times.Once());
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void CustomDeviceModelsCanBeUpserted()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { Id = "id", ETag = "oldEtag" };
+            var updatedDeviceModel = new DeviceModel  { Id = "id", ETag = "newETag" };
+
+            this.SaveDeviceModelInStorage(deviceModel);
+            this.UpdateDeviceModelInStorage(updatedDeviceModel);
+
+            // Act
+            this.target.UpsertAsync(deviceModel).Wait();
+
+            // Assert
+            this.storage.Verify(x => x.UpdateAsync(
+                STORAGE_COLLECTION,
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                "oldEtag"), Times.Once());
+
+            Assert.Equal(updatedDeviceModel.Id, deviceModel.Id);
+            Assert.Equal(updatedDeviceModel.ETag, deviceModel.ETag);
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void ItCreatesDeviceModelWhenDeviceModelNotFoundInUpserting()
+        {
+            // Arrange
+            var deviceModel = new DeviceModel { Id = "id", ETag = "Etag" };
+
+            this.ThereIsNoCustomDeviceModelByIdInStorage(deviceModel.Id);
+            this.UpdateDeviceModelInStorage(deviceModel);
+
+            // Act
+            this.target.UpsertAsync(deviceModel).Wait(TimeSpan.FromSeconds(30));
+
+            // Assert
+            this.storage.Verify(x => x.UpdateAsync(
+                STORAGE_COLLECTION,
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                null));
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void ItThrowsConflictingResourceExceptionIfEtagDoesNotMatchInUpserting()
+        {
+            // Arrange
+            var deviceModelInStorage = new DeviceModel { ETag = "ETag" };
+            var deviceModel = new DeviceModel { ETag = "not-matching-Etag" };
+            
+            this.UpdateDeviceModelInStorage(deviceModelInStorage);
+
+            // Act & Assert
+            Assert.ThrowsAsync<ConflictingResourceException>(() => this.target.UpsertAsync(deviceModel));
+        }
+
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void ItThrowsExceptionWhenInsertDeviceModelFailed()
+        {
+            // Arrange
+            this.SetupStorageToThrowException();
+
+            // Act & Assert
+            Assert.ThrowsAsync<Exception>(() => this.target.InsertAsync(It.IsAny<DeviceModel>()));
```

Comment:
when invoking actions we shouldn't use `It.Is...`. I think this should be InsertAsync( a valid model ). The model should be valid to be sure the exception doesn't occur due to the input

Reply:
Done.